contrib/test/integration/swap.yml

```diff
@@ -0,0 +1,39 @@
+---
```

Comment:
Wouldn't this be better written as a module? Imperative logic to achieve idempotent action as a series of tasks seems really weird.

Reply:
Ansible YAML is a guaranteed to be an understood language in this context.  Unless it was some super-complicated/intricate thing, it's often more compact and easier to validate using existing, proven modules.

Reply:
In terms of theory I agree with you Steve.  The issue here is of practical application.  Remember, our "audience" here is a bunch of "go" nerds: Barely just having stepped out of their Ansible-training-pants 😆 

This is something @edsantiago yells at me about (stomping his feet) constantly.  Balancing brush-stroke-width, for the benefit of *future* maintainers, vs ourselves.  Worse, in my experience it gets MORE difficult, the deeper you understand the topic - possibly because the [more precise tools, become easier to grasp](https://www.brainyquote.com/quotes/quotes/p/pablopicas385182.html).

If this were more purely an IT automation project, and we could afford (in time and people) to write and maintain Ansible modules with unittests.  Then sure, the theory-vs-practice gap could be much narrower.  In this case, in this repo, I think painting with a wider brush is better.  It stands a better chance of being understood and maintained by !us (if/when necessary).

Reply:
> Barely just having stepped out of their Ansible-training-pants 😆

Therefore, to get them on a good start, they should see that the _correct_ way to write maintainable and safe Ansible playbooks is by encoding idempotent logic into modules.

> Balancing brush-stroke-width, for the benefit of future maintainers, vs ourselves. 

This absolutely makes _YOUR_ life easier, too. Do you know the implementation details of every imperative construct that exists in Ansible and how they interact with each other? Do  you know which interactions are bound by an SLA and have well-defined output? Do you know how future versions of the constructs will behave? Do you know in which situations the constructs will act unlike their counterparts in a real procedural language? Do you have the time to remember this every time you work with this playbook? Do you have the mental energy to explain this to new members of the team when they see it for the first time? Are you going to maintain this playbook? Will you need to care for it as it grows organically?

> If this were more purely an IT automation project, and we could afford (in time and people) to write and maintain Ansible modules with unittests. 

I feel like the Ansible module would contain about the same number of lines, it would just be in Python. I do not feel like there is ever a strong argument that unit tests are too much effort to maintain. That's a fairly short-sighted outlook.

I don't mind if you take the current approach, since I will not be maintaining the playbook, but it is making use of -- in my opinion -- the worst anti-pattern that exists for Ansible. The "effort" involved in writing this in Python is most likely less than the effort we went to in having this conversation. The benefits are staggering. The costs are painful.

Reply:
> it would just be in Python

Precisely, and the pillar of my argument...Ansible is the common-denominator:  Few people here know python, not all know Go, and most hate bash.  Everyone speaks at least a little Ansible.  So we compromise out of the necessity to progress CRI-O forward while keeping the testing stuff accessible to the many, instead of the few.

Reply:
Nobody speaks Ansible. I'm not trying to be silly here but seriously -- look at the implementation for the imperative constructs in Ansible. They are NOT a language and o NOT work the way you would expect. Nobody has the mental energy to figure it out. 

If you don't trust me -- look into it!!! When can you mix a `block` with a `when`? Can you add a `with_items` for a `block`? Can you `with_items` over a `role`? Like, it's madness. It has no logic to it. It's just however the team decided to build the supporting Python in Ansible. I don't claim to be the most amazing Ansible dev in the world but I've been around the block.... and I got bitten by the fact that if you do:

```yaml
---
- block:
  - whatever: {{ item }}
    with_items:
      - item
    when: something
  when: something_else
```

You are NOT getting:

```python
if something_else:
    for item in items:
        if something:
            whatever(item)
```

Instead, you get:

```python
for item in items:
    if something and something_else:
        whatever(item)
```

Did YOU know that? Do you expect new devs to Ansible to figure that one out? Using a real language is infinitely easier.

Reply:
Actually, digging into that example even more ... let's talk just about the `with_items` and `when`...

```yaml
---
- whatever: {{ item }}
  with_items:
    - item
  when: something
```

You are NOT getting:

```python
if something:
    for item in items:
        whatever(item)
```

Instead, you get:

```python
for item in items:
    if something:
        whatever(item)
```

How do you know? Just looking at the YAML, without prior knowledge, how can you even figure it out? You can't. This is what I mean. This bites every single person using Ansible this way. The module -- in Python -- is where you should place your complicated logic to achieve idempotency. The Playbook -- in YAML -- should be a list of  things you would like to see exist.

Reply:
Steve, I gotcha, I'm with ya, I agree (and I did know that about when folding).  Yet the language-barrier is still real.  Folks are coming up to speed on the Ansible stuff...slowly...but it's happening.  Keeping testing stuff in once "language" like this means (eventually) you, I, and ed won't be pinged every time there's a change - they can do it themselves 😀

Having testing magic embedded in Python is just as good as Fortran or Go (I'm assuming you don't know the either).  It doesn't matter how well it works, very few people (here) would be able to maintain it.   Bash is about the only other common "language" here, and I don't see anybody jumping up to manipulate JSON from bash 😖   Heck, even YAML in bash is challenging.

(For those following along at home:  Ansible expects "action" modules to return JSON on stdout - matter what language they're written in).

Reply:
I don't follow. New dev joins the team and has never seen Ansible or otherwise in their life. With the current approach:

 1. determine what an Ansible playbook is
 2. determine what an Ansible play is
 3. determine what an Ansible task is
 4. spend a lot of time understanding all of the nuances of how Ansible imperative statements work and having trouble reconciling that functionality with how normal languages work

The flow with a module:

 1. steps 1-3 from above
 2. read Python, which is well-formed, easy to parse and generally understandable

Reply:
Point being -- understanding the core concepts of Ansible is basically orthogonal with the imperative statements/etc.

Reply:
Speaking from the sidelines... @stevekuznetsov , many thanks for your encouragement on this topic. It seems pretty clear that modules are the right approach for some of the gripes I've long had about Ansible. However: the barrier to entry is high. Steep learning curve, plus nontrivial effort to write the module plus tests plus clear documentation. I think it just ain't gonna happen for this PR.

I commit to learning more about modules. I will find time to play with them, write a scratch module, try to fit them into my playbooks. I will do so over weeks and months, and in so doing will make better use of Ansible, and I will appreciate you for the motivation to do so. I imagine Chris will do so as well. But for this PR, I think what Chris has is Good Enough.

Reply:
@mrunalp @rhatdan @TomSweeneyRedHat (NOT urgent)...This PR is ready to go with @edsantiago blessing however...there's an interesting philosophical debate (above).  I'm okay with merging the PR as-is, and resolving the design-issue secondarily (in another PR).  I'll send out a mass-mail to y'all (and copy @stevekuznetsov ) so we can continue the discussion there...

Reply:
@edsantiago @cevich Sounds good to me. Y'all should reach out to Scott Dodson (he sits really close to you, @cevich ). His team has been working through a lot of these similar issues as they manage a large and quickly moving Ansible codebase.

Reply:
Let's merge this and continue the conversation later :) our ansible stuff is too tiny to even debate on something like that here ha